test/controlplane/node/localnode.go

```diff
@@ -0,0 +1,166 @@
+		},
+		Status: corev1.NodeStatus{
+			Conditions: []corev1.NodeCondition{},
+			Addresses: []corev1.NodeAddress{
+				{Type: corev1.NodeInternalIP, Address: "10.0.0.1"},
+				{Type: corev1.NodeExternalIP, Address: "20.0.0.2"},
+				{Type: corev1.NodeHostName, Address: "minimal"},
+			},
+		},
+	}
+)
+
+// errorer implements the interface required by 'assert' to gather
+// the assertion errors.
+type errorer struct {
+	err error
+}
+
+func (e *errorer) Errorf(format string, args ...interface{}) {
+	e.err = multierr.Append(e.err, fmt.Errorf(format, args...))
+}
+
+func validateLocalNode(cs client.Clientset, lns *node.LocalNodeStore) error {
+	// Validate that eventually the local node information is fully populated.
+
+	// The initial assertions should still hold.
+	if err := validateLocalNodeInit(lns); err != nil {
+		return fmt.Errorf("validateLocalNode: %w", err)
+	}
+
+	errs := &errorer{}
+	node := lns.Get()
+
+	// PodCIDR has been populated from the node object.
+	assert.Equal(errs, podCIDR.String(), node.IPv4AllocCIDR.String())
+
+	// HealthIP has been allocated.
+	assert.NotEmpty(errs, node.IPv4HealthIP)
+	// CiliumNode object has been created and populated correctly
+	// and reflects the state of the local node.
+	ciliumNodes, err := cs.CiliumV2().CiliumNodes().List(context.TODO(), metav1.ListOptions{})
+	assert.NoError(errs, err)
+	assert.Len(errs, ciliumNodes.Items, 1)
+
+	ciliumNode := ciliumNodes.Items[0]
+
+	if assert.NotEmpty(errs, ciliumNode.OwnerReferences) {
+		// CiliumNode should have owner reference to Node
+		assert.Equal(errs, localNodeObject.UID, ciliumNode.OwnerReferences[0].UID)
+	}
+
+	parsedCiliumNode := nodeTypes.ParseCiliumNode(&ciliumNode)
+	assert.Equal(errs, node.IPv4HealthIP, parsedCiliumNode.IPv4HealthIP, "CiliumNode HealthIP")
+	assert.Equal(errs, node.IPAddresses, parsedCiliumNode.IPAddresses, "CiliumNode IPAddresses")
+	assert.Equal(errs, node.Labels, parsedCiliumNode.Labels, "CiliumNode Labels")
+
+	// FIXME annotations are lost due to fillLocalNode() not setting them.
+	//assert.Equal(errs, node.Annotations, parsedCiliumNode.Annotations, "CiliumNode Annotations")
+
+	if errs.err != nil {
+		return fmt.Errorf("validateLocalNode: %w", errs.err)
+	}
+	return nil
+}
+
+func validateLocalNodeInit(lns *node.LocalNodeStore) error {
+	// Validate that after LocalNodeStore has started it has been partially populated.
+	// This is called before Daemon is started.
+
+	errs := &errorer{}
+	node := lns.Get()
+
+	// These things we expect to be populated right after
+	// LocalNodeStore has started:
+	assert.Equal(errs, localNodeObject.Name, node.Name)
+	assert.Equal(errs, "10.0.0.1", node.GetNodeIP(false).String())
+	assert.Equal(errs, "20.0.0.2", node.GetExternalIP(false).String())
+	assert.Contains(errs, node.Labels, "foo")
+	assert.Contains(errs, node.Annotations, "baz")
+
+	if errs.err != nil {
+		return fmt.Errorf("validateLocalNodeInit: %w", errs.err)
+	}
+	return nil
+}
+
+func init() {
```

Comment:
TIL another wonderful testing pattern xD

Reply:
Yeah, I surprised myself a bit that I had done this... should refactor this to just explicitly compose the test suite.